### PR TITLE
config: refactor configure for MPI_Aint and MPI_Offset

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description

* `MPI_AINT_FMT_..._SPEC` does not belong in `mpi.h`. They are moved to `mpiimpl.h` in this PR.

* In principle, MPI_Aint and MPI_Offset are either 4 bytes or 8 bytes. `configure` should just probe and define the sizes -- SIZEOF_MPI_AINT and SIZEOF_MPI_OFFSET. The reset of the logic should be carried out at the MPIR-layer using C preprocessor.

* Historically we let ROMIO configure the "offset" type and MPICH inherits it. It is more sensible to make it the other way.

Reference: it was originated in fixing a few romio warnings (PR #1811 and PR #1812).

## Expected Performance Changes

It should not affect runtime behavior. 
It may simplify the build time logic.

## Known Issues

There maybe legacy issues (such as third party usage of ROMIO).

## Author Checklist
* [x] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [x] Passes tests (included warning check)
* [x] Confirm whitespace/style checkers are happy (or has a good reason for being bad)
* [ ] Commits are self-contained and do not do two things at once
* [x] Remove xfail from the test suite when fixing a test
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/quick
skip_test=true

**Comment by hzhou** [comment]
test:mpich/quick
skip_test=true

**Comment by hzhou** [comment]
test:mpich/quick
skip_test=true

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

**Comment by hzhou** [comment]
test:mpich/quick

**Comment by hzhou** [comment]
test:mpich/quick

**Comment by hzhou** [comment]
@roblatham00 This PR is ready for your review. Let me know if you would like me to split this PR into a few smaller ones.

**Comment by pavanbalaji** [comment]
Don’t forget that MPI_Aint needs to be size of void*.  Otherwise attributes will stop working and RMA will have a bunch of warnings.  The BG/P days were a nightmare when they didn’t match.  We should try to not repeat them.

**Comment by hzhou** [comment]
> 
> 
> Don’t forget that MPI_Aint needs to be size of void*. Otherwise, attributes will stop working and RMA will have a bunch of warnings. The BG/P days were a nightmare when they didn’t match. We should try to not repeat them.

If `MPI_Aint` need to be exactly the size of `void *`, the code can be simplified even more. Could you confirm that? Reading the code I thought there are cases that `MPI_Aint` need be defined separately.

The current configure is made much more complex by interweaving the FORTRAN and datatype logic with the core feature detectiion. I would very much like to split out the FORTRAN config into a `.m4` or `subconfgiure.m4` and split the datatype handler logic out as well. What do you think?

**Comment by pavanbalaji** [comment]
The MPI standard has been vague on that question.  One one hand, the MPI standard says that `MPI_Aint` should be large enough to fit a `void *` (so >= `sizeof(void*)`), but on the other hand, it allows users to set `MPI_Aint` attributes as `void *` (so <= `sizeof(void*)`).

On IBM BG/P, we setup a model where `MPI_Aint` > `sizeof(void *)`, but that didn't go well with users.  We had done this because BG/P had a 32-bit `void *`, but a 64-bit filesystem, so we wanted MPI I/O be to be able to use that file system.  It was a mess.

So: given all of that background, in MPICH, I recommend that we always have `MPI_Aint` be the same size as `void *` (so basically `ssize_t`).

**Comment by pavanbalaji** [comment]
Ideally, yes, but the standard uses `MPI_Aint` in a bunch of places.

File: configure.ac [hunk 1]
```diff
@@ -4812,66 +4761,65 @@ if test "$with_aint_size" -gt 0 ; then
         AC_MSG_RESULT([Overriding MPI_Aint to be $aint_size bytes])
    fi
```
Comment (hzhou) [inline configure.ac]: @pavanbalaji Should we remove the option `--with-aint-size`?

Comment (pavanbalaji) [reply configure.ac]: I recommend that.  @roblatham00 what do you think?

**Comment by raffenet** [comment]
> So: given all of that background, in MPICH, I recommend that we always have `MPI_Aint` be the same size as `void *` (so basically `ssize_t`).

A couple pedantic notes:

1. `ssize_t` is defined by POSIX, not C.
2. `ssize_t` is not guaranteed to be the same size as a `void *`. The requirement is that it can represent a implementation defined range of numbers. See https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_types.h.html.
> The type ssize_t shall be capable of storing values at least in the range [-1, {SSIZE_MAX}].

**Comment by pavanbalaji** [comment]
`intptr_t` might be a better type?

**Comment by raffenet** [comment]
`intptr_t` can definitely hold a pointer value. I believe there is still some wiggle room in the definition that could allow the size to differ.

**Comment by hzhou** [comment]
I like the use of `MPI_Aint` instead of `ssize_t` or `size_t`. For one, our code base has been using `MPI_Aint` so there is no culture shift. For two, having a wrapper type gives us the best flexibility. For example, we could `typedef size_t MPI_Aint;`or make it `ssize_t`, or define our own type easily without code revamp. `size_t` after all is at the mercy of the compiler and we may have a future/legacy system that does not fit into standard. 

Having `size_t` sprinkled into our code base -- the inconsistency -- on the other hand, is not good.

**Comment by hzhou** [comment]
Strictly speaking, when `MPI_Aint` and `void *` are the same size, the signed type may not be able to hold the value of an unsigned type. Is this something we should worry?

**Comment by raffenet** [comment]
I previously suggested that `MPI_Aint` was not an ideal choice when referring to count arguments internally in MPICH https://github.com/pmodels/mpich/issues/3365. I still believe this is true, as negative counts (i.e. not displacements) are not allowed and `size_t` is an appropriate C language type to represent a number of bytes to be read/written/copied.

**Comment by hzhou** [comment]
> I previously suggested that `MPI_Aint` was not an ideal choice when referring to count arguments internally in MPICH #3365. I still believe this is true, as negative counts (i.e. not displacements) are not allowed and `size_t` is an appropriate C language type to represent a number of bytes to be read/written/copied.

So here is the new discussion: can we make `MPI_Aint` unsigned? FORTRAN will still use it as signed due to the limit of the language, but as long as FORTRAN code does not make it negative -- which we could claim that is a user error -- the binding interface will hold.

**Comment by hzhou** [comment]
This PR is superseded  by PR #3819. It may be revived if we decide to split parts of PR #3819 back here.

**Pull request closed without merging**